From 7907932e62e493ad13aabe50d37980274bda554b Mon Sep 17 00:00:00 2001
From: Murala S Swamy <ssmurala@habana.ai>
Date: Tue, 05 Nov 2024 07:01:09 +0200
Subject: [PATCH] [FSW-11346] fs1: add secondary image check

With this patch added support to check efuse bit
for secondary image enabled status. Used this
check in boot flow to skip all seconday image
related code executions.

Change-Id: Ie6e635970d6a7b1be7414c1b8fe12a397ccdfc04
---

diff --git a/arch/fs1/arc/common.c b/arch/fs1/arc/common.c
index 2430f6b..faa4c52 100644
--- a/arch/fs1/arc/common.c
+++ b/arch/fs1/arc/common.c
@@ -526,3 +526,8 @@
 {
 	regmap_writel(FULL_ADDR_MSB, IO0_H0_MNG_ARC(mmARC_GP_MSB_ADDR_0));
 }
+
+bool boot_rom_second_image_enabled(void)
+{
+	return ((boot_rom_global.config_fuse & EFUSE_SYSCFG_DIS_SECONDARY_IMG) == 0);
+}
diff --git a/common/boot_flow.c b/common/boot_flow.c
index 7762995..c441acd 100644
--- a/common/boot_flow.c
+++ b/common/boot_flow.c
@@ -76,8 +76,9 @@
 	 * All manifest switches should be tracked through
 	 * single source file boot_flow.c
 	 */
-	/* TODO: Add check if secondary image is supported */
-	if (!!boot_rom_global.fw_upd_switch_manifest) {
+	/* TODO: If 2ndary image disable, we will not support fw update ?*/
+	if (boot_rom_second_image_enabled() &&
+		(!!boot_rom_global.fw_upd_switch_manifest)) {
 		primary_manifest_idx = HL_FW_SPI_MANIFEST_1;
 		img_sts_base = PROGRESS_BRINGUP_PCI_SEC;
 	}
@@ -93,8 +94,10 @@
 		primary_ppboot_prefetch_failed = true;
 	}
 
+	if (!boot_rom_second_image_enabled())
+		return BOOT_ROM_FAIL;
+
 	if (!secondary_ppboot_prefetch_started && !secondary_ppboot_prefetch_failed) {
-		/* TODO: Add check if secondary image is supported */
 		curr_manifest_idx = (primary_manifest_idx + 1) % HL_FW_SPI_MANIFEST_MAX_NUM;
 		ret = boot_flow_prefetch_img_sub(curr_manifest_idx, img_type, &desc, img_sts_base);
 		if (!ret) {
@@ -184,7 +187,8 @@
 			primary_manifest_valid = true;
 	}
 
-	/* TODO: Add check for secondary image disabled */
+	if (!boot_rom_second_image_enabled())
+		return;
 
 	curr_manifest_idx = (primary_manifest_idx + 1) % HL_FW_SPI_MANIFEST_MAX_NUM;
 	if (boot_rom_boot_desc_init_done(curr_manifest_idx)) {
diff --git a/common/common.h b/common/common.h
index 85de629..b38dc2a 100644
--- a/common/common.h
+++ b/common/common.h
@@ -40,5 +40,6 @@
 void boot_rom_config_arc_regions(void);
 void boot_rom_init_slave_spi_efuse(void);
 void boot_rom_wait_for_spi_fuse_data(void);
+bool boot_rom_second_image_enabled(void);
 
 #endif /* _COMMON_H_ */
